### PR TITLE
Add flag tests

### LINKED ISSUE #132: Add tests for output of each flag usage and combination
This would probably be the MOST foolproof way of ensuring that everything is working as expected. Earlier, reviewing PRs used to be easier by looking at the Travis CI build logs. However, with increasing number of flag options, it becomes tougher and tougher to manually keep checking the changes. Hence, the need for **ACTUAL** tests (rspec) arises.

- Relevant Issues : (none)
- Relevant PRs : (none)
- Type of issue :
  - [ ] Installation
  - [ ] Font-related
  - [ ] Feature request
  - [ ] Bug in existing feature
  - [x] Developer mode : Code quality / Tests / Documentation

### PR DESCRIPTION
**Pull request by henrebotha**
### Description

This PR adds tests for the presence & absence of each individual command line flag. It partly resolves #132.

- Relevant Issues : #132 
- Relevant PRs : (none)
- Type of change :
  - [ ] New feature
  - [ ] Bug fix for existing feature
  - [ ] Code quality improvement
  - [x] Addition or Improvement of tests
  - [ ] Addition or Improvement of documentation

### CODE REVIEW

**Comment by henrebotha** [comment]
@athityakumar I'm not sure why the CI build failed, it's complaining about env variables. Any thoughts?

File: README.md [hunk 1]
```diff
@@ -55,7 +55,7 @@ A Ruby script that colorizes the `ls` output with color and icons. Here are the
 
   ![Usage #9](readme/usage9.png)
 
-- Additional flags: `--almost_all` or `-A`, `--long` or `-l`, `--tree` or `-t`
+- Additional flags: `--almost-all` or `-A`, `--long` or `-l`, `--tree` or `-t`
```
Comment (athityakumar) [inline README.md]: Good catch. 😄

**Comment by athityakumar** [comment]
@henrebotha - Travis CI is failing, definitely because of icons mismatch. You're checking only for text, whereas actual output contains the icons as well. You can use the `match` matcher in rspec to match output with couple of regexps like [this](https://relishapp.com/rspec/rspec-expectations/v/2-2/docs/matchers/match-matcher).

Even better solution would be to have an exact match (with icons) rather than just regexp match. You can also write custom matcher like `line_contains` in case of multi-line outputs.

**Comment by henrebotha** [comment]
Dope, will do

File: spec/colorls_spec.rb [hunk 2]
```diff
@@ -1,7 +1,85 @@
 require 'spec_helper'
 
 RSpec.describe ColorLS do
+  FIXTURES = 'spec/fixtures'.freeze
```
Comment (athityakumar) [inline spec/colorls_spec.rb]: Add a global `subject { ColorLS::Flags.new(flags).process }` and define individual context-flags with `let`.

**Comment by henrebotha** [comment]
If I apply `rspec-its` as you requested, can I do multiple assertions per `context`? E.g.

```ruby
context 'checking message' do
  subject { something }

  it { is_expected.to output(result).to_stdout }
  it { is_expected.to_not output(result_alt).to_stdout }
end
```

**Comment by athityakumar** [comment]
@henrebotha - Yup. Infact, the best part is - you can set the subject as global (next to `FIXTURES`).

```ruby
subject { ColorLS::Flags.new(*args).process }
```

```ruby
context 'check for flag -A' do
  let(:args) { ['-A', FIXTURES] }

  it { is_expected.to output(result).to_stdout }
  it { is_expected.to_not output(result_alt).to_stdout }
end
```

**Comment by athityakumar** [comment]
BTW, can you shift these testing codes into a new file (`spec/flags_spec.rb`)? That way, it can be -

```ruby
RSpec.describe ColorLS::Flags do
  subject { described_class.new(*args).process }

  ...
end
```

**Comment by henrebotha** [comment]
Ah yeah, will do

**Comment by henrebotha** [comment]
I'm struggling to capture the stdout using `rspec-its`. I get this style of error:

```
expected block to not output /foo/ to stdout, but was not a block
       Diff:
       @@ -1,2 +1,2 @@
       -/foo/
       +""
```

Googling for `rspec-its` and stdout doesn't yield much.

**Comment by henrebotha** [comment]
I guess `ColorLS::Core#ls` is side-effecting... so we can't directly capture the output as a string. Hmm...

**Comment by athityakumar** [comment]
If you can push your changes, It'd be easier to review. 😄

**Comment by henrebotha** [comment]
Lol, I don't like forcing people to review broken code! 😂  But I'll do that

**Comment by henrebotha** [comment]
Under "RSpec < 3.0 (and others)", [this answer](https://stackoverflow.com/a/16507680/1966418) presents a way to wrap the call so as to capture stdout. I could use something like it? Unless there's a "proper" way to make rspec-its play nice with stdout...

**Comment by athityakumar** [comment]
Our rspec version on CI is 3.7, so that shouldn't be the case. Try 

```ruby
before { described_class.new(*args).process }

subject { STDOUT }
```

**Comment by henrebotha** [comment]
`subject { STDOUT }` gives `"#<IO:<STDOUT>>"`, not the actual contents. Calling `to_s` on it doesn't help either.

> Our rspec version on CI is 3.7, so that shouldn't be the case.

I'm aware, but post-3.0 tutorials for how to match against stdout invariably use `output`, which doesn't seem to be working with rspec-its. I tried the method given in that SO answer and it works perfectly, but I'm willing to try and get the tests to work more like how you're describing.

**Comment by henrebotha** [comment]
Actually, to be more clear: doing `subject { STDOUT }` gives me the same error as before. When I change the expectations from `output` to `match`, it tries to match against `"#<IO:<STDOUT>>"`.

File: spec/flags_spec.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  context 'with --almost-all flag' do
+    let(:args) { ['--almost-all', FIXTURES] }
+
+    it { is_expected.to match(/\.hidden-file/) } # displays hidden files
+  end
+
+  context 'with --tree flag' do
+    let(:args) { ['--tree', FIXTURES] }
+
+    it { is_expected.to match(/├──/) } # displays file hierarchy
```
Comment (athityakumar) [inline spec/flags_spec.rb]: All these comments like `# displays file hierarchy` are supposed to be in the context message too. Like `context 'with --tree flag displays file hierarchy' do`.

File: spec/flags_spec.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  context 'with no flags' do
+    let(:args) { [FIXTURES] }
+
+    it { is_expected.to_not match(/((r|-).*(w|-).*(x|-).*){3}/) } # does not list file info
+    it { is_expected.to_not match(/\.hidden-file/) } # does not list hidden files
+    it { is_expected.to_not match(/Found \d+ contents/) } # does not show a report
+    it { is_expected.to match(/a-file.+symlinks.+z-file/) } # sorts all results alphabetically
+    it { is_expected.to match(/a-file.+symlinks/) } # displays dirs & files
```
Comment (athityakumar) [inline spec/flags_spec.rb]: Lines 23 & 24: Can they be split into individual matches of `a-file`, `symlinks` and `z-file`?

Comment (henrebotha) [reply spec/flags_spec.rb]: Oh good catch, I didn't think to reduce the redundancy here when I moved stuff around.

File: spec/flags_spec.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    $stdout = old
+  end
+
+  context 'with no flags' do
+    let(:args) { [FIXTURES] }
+
+    it { is_expected.to_not match(/((r|-).*(w|-).*(x|-).*){3}/) } # does not list file info
+    it { is_expected.to_not match(/\.hidden-file/) } # does not display hidden files
+    it { is_expected.to_not match(/Found \d+ contents/) } # does not show a report
+    it { is_expected.to match(/a-file.+symlinks.+z-file/) } # displays dirs & files alphabetically
```
Comment (athityakumar) [inline spec/flags_spec.rb]: Reduce redundancy in this line too.

Comment (henrebotha) [reply spec/flags_spec.rb]: It doesn't seem redundant to me. It's checking that files and dirs are mixed together & sorted alphabetically. In other words, it should _not_ match this order...

```
a-file
z-file
symlinks
```

...or this one:

```
symlinks
a-file
z-file
```

Comment (athityakumar) [reply spec/flags_spec.rb]: Hmm, this makes sense. Why does it show a different order in Travis CI build?

**Comment by athityakumar** [comment]
Rebase with master and fix the Rubocop issue please. I'll see what's going on with the Rspec issue on travis CI builds.

**Comment by henrebotha** [comment]
I'll try to fix the rubocop errors, but just noting that they weren't created by my code.

**Comment by athityakumar** [comment]
Yes, I know. It was one of the merges made on the master, but it reflects on subsequent PRs too.

**Comment by henrebotha** [comment]
I don't understand `lib/tab_complete.sh`. It looks like Ruby mixed with shell. Not sure how to address the linter errors.

**Comment by athityakumar** [comment]
@henrebotha - It's just a plain shell script with no Ruby. Exclude it from Rubocop's vision (check .rubocop.yml file).

File: spec/flags_spec.rb (hunk 3 above)
Comment (henrebotha) [reply spec/flags_spec.rb]: I think the way it is now, the error messages read very nicely:

```
Failures:

  1) ColorLS::Flags with no flags should match /\.hidden-file/
     Failure/Error: it { is_expected.to match(/\.hidden-file/) } # does not display hidden files
  2) ColorLS::Flags with no flags should match /Found \d+ contents/
     Failure/Error: it { is_expected.to match(/Found \d+ contents/) } # does not show a report
```

If I change it the way you're asking, we lose this. And we also lose the ability to do multiple assertions per context - as it is, `context 'with no flags'` does all the assertions for calling colorls with no flags. Your original goal with #132 was, of course, tests for each possible combination of flags; if we can't group those, then there's a combinatorial explosion of tests that need to be written (to the point where I think the tests lose some of their usefulness, perhaps).

If you're happy with all of that, then I'll go ahead and make the change.

File: spec/flags_spec.rb (hunk 5 above)
Comment (athityakumar) [reply spec/flags_spec.rb]: How about adding just this specific `it` within a pending context? Opening an issue to work on this in a different PR, would be better. Have this as pending and let's merge the rest of the specs.

File: spec/flags_spec.rb (hunk 3 above)
Comment (athityakumar) [reply spec/flags_spec.rb]: Makes sense, let's skip this.

File: spec/flags_spec.rb (hunk 5 above)
Comment (henrebotha) [reply spec/flags_spec.rb]: Alright, how shall I mark it as todo?

Comment (henrebotha) [reply spec/flags_spec.rb]: NVM, I understand

**Comment by henrebotha** [comment]
@athityakumar Is this alright?

**Pull request merged**